documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1182 @@
+- Every Dimension should have the relevant number of properties (X and Y for Vector2, and X,Y,Z,W,V for Vector5) and relevant unit vectors
+- Constructors for higher dimensions should include lower dimension variants that use the lower dimensions for their specific components (vector2 -> X,Y)
+
+```cs
+public readonly record struct Vector2I<TScalar> : IEquatable<Vector2I<TScalar>>, 
+                           IEquatable<Vector2>,
+                           IBinaryIntegerVector<Vector2I<TScalar>, TScalar>,
+                           IVector2<Vector2I<TScalar>, TScalar>
+                           where TScalar : IBinaryInteger<TScalar>
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2I<TScalar> UnitX { get; }
+    public static Vector2I<TScalar> UnitY { get; }
+
+    public static Vector2I<TScalar> Zero { get; }
+    public static Vector2I<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2I(TScalar x, TScalar y);
+
+    public static Vector2I<TScalar> operator +(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator -(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator *(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator /(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator %(Vector2I<TScalar> left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator *(Vector2I<TScalar> left, TScalar right);
+    public static Vector2I<TScalar> operator /(Vector2I<TScalar> left, TScalar right);
+    public static Vector2I<TScalar> operator %(Vector2I<TScalar> left, TScalar right);
+
+    public static Vector2I<TScalar> operator *(TScalar left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator +(Vector2I<TScalar> value);
+    public static Vector2I<TScalar> operator -(Vector2I<TScalar> value);
+
+    public static Vector2I<TScalar> operator &(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator |(Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> operator ^(Vector2I<TScalar> left, Vector2I<TScalar> right);
+
+    public static Vector2I<TScalar> operator ~(Vector2I<TScalar> value);
+
+    public TScalar LengthSquared { get; }
+
+    public Vector2I<TScalar> Dot(Vector2I<TScalar> other);
+    public TScalar DistanceSquared(Vector2I<TScalar> other);
+    public Vector2I<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
```

Comment:
These methods will be provided via extension methods to reduce minimum JIT generic specialization code size. They will have explicit implementations of the IVector methods that simply delegate to the extension methods.

Reply:
This is an implementation detail, which isn't meant to be part of the proposal. Whether they are done via extension methods or not shouldn't matter just that they exist on this type.

Reply:
I’m not sure I agree with the assessment on JIT code size here, but concur that this is an implementation detail.

Reply:
> I’m not sure I agree with the assessment on JIT code size here

I gave some descriptions in Discord previously. But it overal impacts trimability and VM overhead due to the forced method table entries per type.

Making it a static method in another class means that you're not taking `this` as an implicit byref, means you can make it an extension method who prefer `x.Dot(y)` over `Vector2.Dot(x, y)`, improve trimming support, reduce VM overhead, etc.

Reply:
-- Notably, it's unclear how these are an implementation detail if they are `public`

Reply:
The part that is an implementation detail is whether these are from extension methods or live in the primary struct. From the user's perspective assuming they both are in the same namespace, it will looks almost the same.

Reply:
True, per the approved SDP we only guarantee source compatibility.